shared/libraries/websocket_streaming/src/websocket_streaming_server.cpp

```diff
@@ -86,4 +92,90 @@ void WebsocketStreamingServer::stopInternal()
+        case CoreEventId::ComponentAdded:
+            componentAdded(sender, eventArgs);
+            break;
+        case CoreEventId::ComponentRemoved:
+            componentRemoved(sender, eventArgs);
+            break;
+        case CoreEventId::ComponentUpdateEnd:
+            componentUpdated(sender);
+            break;
+        default:
+            break;
+    }
+}
+
+void WebsocketStreamingServer::componentAdded(ComponentPtr& /*sender*/, CoreEventArgsPtr& eventArgs)
+{
+    ComponentPtr addedComponent = eventArgs.getParameters().get("Component");
+
+    auto deviceGlobalId = device.getGlobalId().toStdString();
+    auto addedComponentGlobalId = addedComponent.getGlobalId().toStdString();
+    if (addedComponentGlobalId.find(deviceGlobalId) != 0)
+        return;
+
+    LOG_I("Added Component: {};", addedComponentGlobalId);
+    addSignalsOfComponent(addedComponent);
+}
+
+void WebsocketStreamingServer::addSignalsOfComponent(ComponentPtr& component)
+{
+    auto signalsToAdd = List<ISignal>();
+
+    if (component.supportsInterface<ISignal>())
+    {
+        LOG_I("Added Signal: {};", component.getGlobalId());
+        signalsToAdd.pushBack(component.asPtr<ISignal>());
+    }
+    else if (component.supportsInterface<IFolder>())
+    {
+        auto nestedComponents = component.asPtr<IFolder>().getItems(search::Recursive(search::Any()));
+        for (const auto& nestedComponent : nestedComponents)
+        {
+            if (nestedComponent.supportsInterface<ISignal>())
+            {
+                LOG_I("Added Signal: {};", nestedComponent.getGlobalId());
+                signalsToAdd.pushBack(nestedComponent.asPtr<ISignal>());
+            }
+        }
+    }
+
+    streamingServer.addSignals(signalsToAdd);
+}
+
+void WebsocketStreamingServer::componentRemoved(ComponentPtr& sender, CoreEventArgsPtr& eventArgs)
+{
+    StringPtr removedComponentLocalId = eventArgs.getParameters().get("Id");
+
+    auto deviceGlobalId = device.getGlobalId().toStdString();
+    auto removedComponentGlobalId =
+        sender.getGlobalId().toStdString() + "/" + removedComponentLocalId.toStdString();
+    if (removedComponentGlobalId.find(deviceGlobalId) != 0)
+         return;
+
+    LOG_I("Component: {}; is removed", removedComponentGlobalId);
+    streamingServer.removeComponentSignals(removedComponentGlobalId);
+}
+
+void WebsocketStreamingServer::componentUpdated(ComponentPtr& updatedComponent)
+{
+    auto deviceGlobalId = device.getGlobalId().toStdString();
+    auto updatedComponentGlobalId = updatedComponent.getGlobalId().toStdString();
+    if (updatedComponentGlobalId.find(deviceGlobalId) != 0)
+        return;
+
+    LOG_I("Component: {}; is updated", updatedComponentGlobalId);
+
+    // remove all registered signal of updated component since those might be modified or removed
+    streamingServer.removeComponentSignals(updatedComponentGlobalId);
+
+    // add updated versions of signals
+    addSignalsOfComponent(updatedComponent);
+}
```

Comment:
This is quite an expensive change - basically we will lose data on every single signal of the component. We also shouldn't really care too much about modifications here - those should be propagated through the data path as descriptor changes. 

We should, however, handle any removals/additions.

Reply:
Fixed